op-deployer/pkg/deployer/verify/constructors.go

```diff
@@ -0,0 +1,53 @@
+package verify
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+)
+
+var constructorArgSlots = map[string]int{
```

Comment:
does this mean it can only verify these contracts and if they change, it will break and these constants will need to be updated?

Reply:
if you have access to the abi, you can grab the constructor's number of arguments, see https://github.com/ethereum-optimism/optimism/blob/35d4fc19575976167c90103c07bb24b976bfe108/packages/contracts-bedrock/snapshots/abi/SequencerFeeVault.json#L21

Reply:
This is based on the Notion discussion [here](https://www.notion.so/op-deployer-contract-verification-1abf153ee16280a28546c8dd768e350b?d=1abf153ee16280a2b4bf001c43d23285&pvs=4#1abf153ee1628092a474dc4989bf4aa8). Net-net, we will have to keep this list up-to-date but we don't have a lot of great options here:

- We can shell out to Forge, which will require us to pull the monorepo at a certain commit (ugly, error prone, requires external tool)
- We can reconstruct the constructor args in a Solidity script (verbose, requires a lot of Solidity code) or Golang (brittle, hard to maintain)
- We can do something like this (brittle, but simple enough to maintain)

This method, while still not great, at least centralizes the changes we have to make for verification to work.

Reply:
Got it. I am sure you have way more context than me on the tradeoffs and the way its being used, just want to share that [w3](https://github.com/lmittmann/w3?tab=readme-ov-file#abi-bindings) makes it really easy to abi encode in Golang directly

Reply:
Thank you for the idea @tynes. We do indeed have access to the abi and can use that to automate how we set the number of constructor args (and get rid of the `constructorArgSlots` map you referenced in this thread). That helps make this code less brittle and will limi maintenance when constructor args change